### DISCUSSION TITLE
Split table migration

### DISCUSSION THREAD

**Discussion by danilotomaccio** [simolus3/drift#2328] [Q&A]
Hello,
I should, in the new version of a database, split a table into 3 tables: two tables should contain the contents of the previous one (half each) and the third should just connect the other two:

# V1
TableA 
|field1|field2|field3|field4|field5|
|-------|-------|-------|-------|-------|

# V2
TableA 
|field1|field2|
|-------|-------|
|From old TableA|From old TableA|

TableB 
|field3|field4|
|-------|-------|
|From old TableA|From old TableA|

TableC 
| referenceToTableA | referenceToTableB |field5|
|-------|-------|-------|
|Ref to id from new TableA|Ref to id from new TableB|From old TableA|

How do I transfer data from the old version to the new one?
Should I do something like this (this is not working)?
```dart
@override
  MigrationStrategy get migration {
    return MigrationStrategy(
      onCreate: (Migrator m) async {
        await m.createAll();
      },
      onUpgrade: (Migrator m, int from, int to) async {
        if (from < 2) {
          final v2TableC = v2.TableC(this);
          final v2TableB = v2.TableB(this);
          final v2TableA = v2.TableA(this);
          final v1TableA = v1.TableA(this);
          await m.createTable(v2TableC);
          await m.createTable(v2TableB);

          await m.alterTable(TableMigration(v2PairingFoods, columnTransformer: {
            v2TableB.field1 : v1TableA.field1,
            v2TableB.field2: v1TableA.field2
          }));
          await m.alterTable(TableMigration(v2TableA));
          await m.alterTable(TableMigration(v2TableC, columnTransformer: {
            v2TableC.referenceToTableA: v2TableA.id,
            v2TableC.referenceToTableB: v2TableB.id,
            v2TableC.field5: v1TableA.field,
          }));
        }
      },
    );
  }
```

**Comment by kuhnroyal** [comment]
This kind of question belongs on stackoverflow, not here 🤨
1. Create new tables
2. Select from old table and insert into new tables
3. Delete old table
4. Write a migration test

**Reply by danilotomaccio** [reply]
It's a drift related question. I read in the documentation, I quote, "If you get stuck along the way, don't hesitate to [open a discussion about it](https://github.com/simolus3/drift/discussions). Also, there's a working example [in the drift repository](https://github.com/simolus3/drift/tree/latest-release/examples/migrations_example)"

**Reply by simolus3** [reply]
Indeed, there is no way to do all this with a simple `TableMigration` - that API can only use columns from the affected table. And there will also be a problem because `v2TableA` will conflict with `v1TableA` in those queries. 

So I recommend these steps, although I haven't tried them:

1. Create new tables under temporary names (at least `v2tableA` will need a temporary name)
2. Select from old table to insert into the new tables
3. Delete old table
4. Rename new tables to their proper names
5. Definitely write a migration test for this because this is complicated ;)

The next drift version will support inserts with select statements as sources from the Dart API, which will make it fairly easy to write this migration:

```dart
        if (from < 2) {
          final v2TableC = v2.TableC(this);
          final v2TableB = v2.TableB(this);
          final v2TableA = v2.TableA(this).alias('temp_a');
          final v1TableA = alias(v1.TableA(this));
      
          await m.createTable(v2TableC);
          await m.createTable(v2TableB);

          await into(v2TableA).insertFromSelect(
            select(v1TableA),
            columns: {
              v2TableA.field1: v1TableA.field1,
              v2TableA.field2: v1TableA.field2,
            },
          );
          await into(v2TableB).insertFromSelect(
            select(v1TableA),
            columns: {
              v2TableA.field3: v1TableA.field3,
              v2TableA.field4: v1TableA.field4,
            },
          );
          // The insert into v2TableC depends on how you define ids, but you can create a select statement that selects ids from the new tables based on fields from the old tables

          await customStatement('DROP TABLE a');
          await customStatement('ALTER TABLE temp_a RENAME TO a');
        }
```

Until then, the best way to implement step 2 is to use custom select statements defining the insert, so

```dart
await customInsert('INSERT INTO tableA (field1, field2) SELECT field1, field2 FROM tableA');
```
